src/backend/distributed/planner/multi_master_planner.c

```diff
@@ -244,6 +250,71 @@ BuildSelectStatement(Query *masterQuery, List *masterTargetList, CustomScan *rem
 }
 
 
+/*
+ * QueryContainsAggregateWithHLL returns true if the query has an hll aggregate
+ * function in it's target list.
+ */
+static bool
+QueryContainsAggregateWithHLL(Query *query)
+{
+	List *varList = NIL;
+	ListCell *varCell = NULL;
+
+	varList = pull_var_clause((Node *) query->targetList, PVC_INCLUDE_AGGREGATES);
+	foreach(varCell, varList)
+	{
+		Var *var = (Var *) lfirst(varCell);
+
+		if (nodeTag(var) == T_Aggref)
+		{
+			Aggref *aggref = (Aggref *) var;
+			int argCount = list_length(aggref->args);
+			Oid hllId = get_extension_oid(HLL_EXTENSION_NAME, false);
+			Oid hllSchemaOid = get_extension_schema(hllId);
+			const char *hllSchemaName = get_namespace_name(hllSchemaOid);
+			Oid addFunctionId = FunctionOid(hllSchemaName, HLL_ADD_AGGREGATE_NAME,
+											argCount);
+			Oid unionFunctionId = FunctionOid(hllSchemaName, HLL_UNION_AGGREGATE_NAME, 1);
+
+			if (aggref->aggfnoid == addFunctionId || aggref->aggfnoid == unionFunctionId)
+			{
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * UseGroupAggregateWithHLL first checks whether the HLL extension is loaded, if
+ * it is not then simply return false. Otherwise, checks whether the client set
+ * the hll.enable_hashagg to false. If it is disabled and the master query contains
+ * hll aggregate function, it returns true.
+ */
+UseGroupAggregateWithHLL(Query * masterQuery)
+{
+	Oid hllId = get_extension_oid(HLL_EXTENSION_NAME, false);
+	const char *gucStrValue = NULL;
+
+	/* If HLL extension is not loaded, return false */
+	if (hllId == InvalidOid)
+	{
+		return false;
+	}
+
+	/* If HLL is loaded but related GUC is not set, return false */
+	gucStrValue = GetConfigOption(HLL_ENABLE_HASH_GUC_NAME, false, false);
```

Comment:
Can we have hll extension defined in a different schema and still has the guc with hll prefix ?

Reply:
Should be fine. GUC prefixes are unrelated to schemas.